terraform workspace fails to plan/apply when workspace already exists

terraform plan (and apply) with the workspace feature seem to fail if the workspace already exists.

Scenario: code on github, existing project with 2 pre-existing terraform workspaces and remote states in s3.

digger.yml:

```
projects:
- name: shared
  branch: /test-branch/
  dir: ./terraform/
  workspace: shared
  terraform_version: v1.2.0
- name: dev
  branch: /test-branch/
  dir: ./terraform/
  workspace: dev
  terraform_version: v1.2.0
```

from cli:

```
[annalisa:.../tf-project/terraform]$ terraform workspace list
* default
  dev
  shared
```

digger output for plan (on pr in my case):

```
2023/06/07 15:21:01 Following projects are impacted by pull request #2
shared
dev
GitHub event processed successfully
GitHub event converted to commands successfully
2023/06/07 15:21:01 Following commands are going to be executed:
project: shared: commands: "digger plan", 
project: dev: commands: "digger plan", 
Lock annalisabazzi/tf-project#shared
2023/06/07 15:21:02 Lock result: true

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding hashicorp/template versions matching "~> 2.2"...
- Finding hashicorp/aws versions matching "~> 4.57"...
- Installing hashicorp/aws v4.67.0...
- Installed hashicorp/aws v4.67.0 (signed by HashiCorp)
- Installing hashicorp/template v2.2.0...
- Installed hashicorp/template v2.2.0 (signed by HashiCorp)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
default
Workspace "shared" already exists
Error: exit status 1
2023/06/07 15:21:14 Failed to run digger plan command. error executing plan: exit status 1
Failed to run commands. failed to run digger plan command. error executing plan: exit status 1
Error: Process completed with exit code 8.
```

Just for extra reference (although i believe it's not relevant), here the github workflow I'm using to test

```
name: DiggerCI

on:
  pull_request:
    branches: [ "main" ]
    types: [ closed, opened, synchronize, reopened ]
  issue_comment:
    types: [created]
    if: contains(github.event.comment.body, 'digger')
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: digger run
        uses: diggerhq/digger@v0.1.23
        with:
          setup-aws: true
          aws-access-key-id: ${{ secrets.TF_AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: ${{ secrets.TF_AWS_SECRET_ACCESS_KEY }}
          aws-region: eu-west-1
        env:
          GITHUB_CONTEXT: ${{ toJson(github) }}
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

--- a/pkg/core/terraform/tf.go
+++ b/pkg/core/terraform/tf.go
@@ -88,13 +88,22 @@ func (tf Terraform) Apply(params []string, plan *string, envs map[string]string)
 	if err != nil {
 		return "", "", err
 	}
-
+	
 	if strings.TrimSpace(workspace) != tf.Workspace {
-		_, _, _, err = tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
+		_, stderr, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
+
 		if err != nil {
-			return "", "", err
-		}
+			if strings.Contains(stderr, "doesn't exist") {
+				_, _, _, err := tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
+				if err != nil {
+					return "", "", err
+				}
+			} else {
+				return "", "", err
+			}
+		} 
 	}
+	
 	params = append(append(append(params, "-input=false"), "-no-color"), "-auto-approve")
 	if plan != nil {
 		params = append(params, *plan)
@@ -128,6 +137,8 @@ func (tf Terraform) runTerraformCommand(command string, envs map[string]string, 
 		fmt.Println("Error:", err)
 	}
 
+	// fmt.Println("Debug stdout:", stdout)
+
 	return stdout.String(), stderr.String(), cmd.ProcessState.ExitCode(), err
 }
 
@@ -152,17 +163,27 @@ func (sw *StdWriter) GetString() string {
 }
 
 func (tf Terraform) Plan(params []string, envs map[string]string) (bool, string, string, error) {
-	workspace, _, _, err := tf.runTerraformCommand("workspace", envs, "show")
-
+	workspaces, _, _, err := tf.runTerraformCommand("workspace", envs, "list")
 	if err != nil {
 		return false, "", "", err
 	}
-	if strings.TrimSpace(workspace) != tf.Workspace {
-		_, _, _, err = tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
+
+	workspaces = strings.Replace(workspaces, "* ", "", -1)
+	workspaces = strings.Replace(workspaces, "\n", ",", -1)
+	workspaces = strings.TrimSpace(workspaces)
+
+	if strings.Contains(workspaces, tf.Workspace) {
+		_, _, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
 		if err != nil {
 			return false, "", "", err
 		}
+	} else {
+		_, _, _, err := tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
+		if err != nil {
+			return false, "", "", err
+		}		
 	}
+
 	params = append(append(params, "-input=false"), "-no-color")
 	stdout, stderr, statusCode, err := tf.runTerraformCommand("plan", envs, params...)
 	if err != nil {

--- a/pkg/core/terraform/tf.go
+++ b/pkg/core/terraform/tf.go
@@ -87,11 +87,7 @@ func (tf Terraform) Apply(params []string, plan *string, envs map[string]string)
 	if err != nil {
 		return "", "", err
 	}
-
-	workspaces = strings.Replace(workspaces, "* ", "", -1)
-	workspaces = strings.Replace(workspaces, "\n", ",", -1)
-	workspaces = strings.TrimSpace(workspaces)
-
+	workspaces = tf.formatTerraformWorkspaces(workspaces)
 	if strings.Contains(workspaces, tf.Workspace) {
 		_, _, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
 		if err != nil {
@@ -103,7 +99,6 @@ func (tf Terraform) Apply(params []string, plan *string, envs map[string]string)
 			return "", "", err
 		}		
 	}
-	
 	params = append(append(append(params, "-input=false"), "-no-color"), "-auto-approve")
 	if plan != nil {
 		params = append(params, *plan)
@@ -137,8 +132,6 @@ func (tf Terraform) runTerraformCommand(command string, envs map[string]string, 
 		fmt.Println("Error:", err)
 	}
 
-	// fmt.Println("Debug stdout:", stdout)
-
 	return stdout.String(), stderr.String(), cmd.ProcessState.ExitCode(), err
 }
 
@@ -162,16 +155,20 @@ func (sw *StdWriter) GetString() string {
 	return s
 }
 
+func (tf Terraform) formatTerraformWorkspaces(list string) (string) {
+
+	list = strings.TrimSpace(list)
+	char_replace := strings.NewReplacer("*", "", "\n", ",", " ", "")
+	list = char_replace.Replace(list)
+	return list
+}
+
 func (tf Terraform) Plan(params []string, envs map[string]string) (bool, string, string, error) {
 	workspaces, _, _, err := tf.runTerraformCommand("workspace", envs, "list")
 	if err != nil {
 		return false, "", "", err
 	}
-
-	workspaces = strings.Replace(workspaces, "* ", "", -1)
-	workspaces = strings.Replace(workspaces, "\n", ",", -1)
-	workspaces = strings.TrimSpace(workspaces)
-
+	workspaces = tf.formatTerraformWorkspaces(workspaces)
 	if strings.Contains(workspaces, tf.Workspace) {
 		_, _, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
 		if err != nil {
@@ -183,7 +180,6 @@ func (tf Terraform) Plan(params []string, envs map[string]string) (bool, string,
 			return false, "", "", err
 		}		
 	}
-
 	params = append(append(params, "-input=false"), "-no-color")
 	stdout, stderr, statusCode, err := tf.runTerraformCommand("plan", envs, params...)
 	if err != nil {

--- a/pkg/core/terraform/tf.go
+++ b/pkg/core/terraform/tf.go
@@ -83,25 +83,25 @@ func (tf Terraform) Init(params []string, envs map[string]string) (string, strin
 }
 
 func (tf Terraform) Apply(params []string, plan *string, envs map[string]string) (string, string, error) {
-	workspace, _, _, err := tf.runTerraformCommand("workspace", envs, "show")
-
+	workspaces, _, _, err := tf.runTerraformCommand("workspace", envs, "list")
 	if err != nil {
 		return "", "", err
 	}
-	
-	if strings.TrimSpace(workspace) != tf.Workspace {
-		_, stderr, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
 
+	workspaces = strings.Replace(workspaces, "* ", "", -1)
+	workspaces = strings.Replace(workspaces, "\n", ",", -1)
+	workspaces = strings.TrimSpace(workspaces)
+
+	if strings.Contains(workspaces, tf.Workspace) {
+		_, _, _, err := tf.runTerraformCommand("workspace", envs, "select", tf.Workspace)
 		if err != nil {
-			if strings.Contains(stderr, "doesn't exist") {
-				_, _, _, err := tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
-				if err != nil {
-					return "", "", err
-				}
-			} else {
-				return "", "", err
-			}
-		} 
+			return "", "", err
+		}
+	} else {
+		_, _, _, err := tf.runTerraformCommand("workspace", envs, "new", tf.Workspace)
+		if err != nil {
+			return "", "", err
+		}		
 	}
 	
 	params = append(append(append(params, "-input=false"), "-no-color"), "-auto-approve")

